### DISCUSSION TITLE
Is it possible to automatically convert a Union type to only one type automatically with pydantic?

### DISCUSSION THREAD

**Discussion by link89** [pydantic/pydantic#4962] [Question]
Given the following data model:
```python

class Demo(BaseModel):
  id: Union[int, str]
  files: Union[str, List[str]]
```
Is there a way to tell `pydantic` to convert id to str type and files to List[str] type automatically, instead of doing this manually every time.

**Comment by samuelcolvin** [comment]
I don't understand the question.

**Reply by link89** [reply]
More example below
```python
obj: Demo = Demo.parse_obj({'id': 1, 'files': '/files/a'})

# when I need to access id or files, I have to convert them manually like below
id = str(obj.id)  # ensure it is str
files = [files] if isinstance(obj.files, str) else files   # ensure it is List[str]

```
Is there a way that I can define some post-process method in the data model, so that when I access `obj.id` it is always `str`, and `obj.files` is always `List[str]`

**Reply by samuelcolvin** [reply]
Use a validator

https://docs.pydantic.dev/usage/validators/

You can probably remove the union.

**Reply by link89** [reply]
I try this but doesn't work, do you know what's wrong with my code

```python
from pydantic import BaseModel, validator
from typing import List

class Demo(BaseModel):
    id: str
    files: List[str]

    @validator('id')
    def id_must_be_str(cls, v):
        if isinstance(v, int):
            v = str(v)
        assert isinstance(v, str)
        return v

    @validator('files')
    def files_must_be_list_of_str(cls, v):
        if isinstance(v, str):
            v = [v]
        assert isinstance(v, list)
        return v


obj = Demo.parse_obj({'id': 1, 'files': '/data/1.txt'})

print(type(obj.id))
print(type(obj.files))
```
The error is
```
pydantic_demo.py", line 23, in <module>
    obj = Demo.parse_obj({'id': 1, 'files': '/data/1.txt'})
  File "pydantic/main.py", line 521, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Demo
files
  value is not a valid list (type=type_error.list)
```

**Reply by link89** [reply]
Now I know what I am looking for is a pre-process hook instead of validator or post-process hook. If pydantic have a hook that allows user to convert values of specific fields before validate them, it will solve my problem.

**Reply by samuelcolvin** [reply]
Read the docs!

You need `pre=True`

**Reply by link89** [reply]
Oh, I see. Thank you for your help.

**Reply by link89** [reply]
By the way, why not consider have a dedicated decorator for pre-process? Honestly I don't expect to find such feature in validator.
